src/test/java/com/google/jenkins/plugins/computeengine/integration/RealJenkinsLogUtil.java

```diff
@@ -0,0 +1,49 @@
+package com.google.jenkins.plugins.computeengine.integration;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.hasProperty;
+import static org.hamcrest.Matchers.not;
+
+import com.google.jenkins.plugins.computeengine.CleanLostNodesWork;
+import hudson.logging.LogRecorder;
+import hudson.logging.LogRecorderManager;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import org.jvnet.hudson.test.JenkinsRule;
+
+/**
+ * Utility class for asserting log messages in Jenkins when using {@code RealJenkinsRule}.
+ * If the test is only using {@code JenkinsRule}, then use the {@code LoggerRule} instead.
+ * Note: Probably it is better to have these kind of utilities in the jenkins-test-harness itself instead, but at least as of now, such utilities are not available.
+ */
+public class RealJenkinsLogUtil {
+
+    public static void setupLogRecorder(JenkinsRule j, String logRecorderName) {
+        LogRecorderManager lrm = j.jenkins.getLog();
+        LogRecorder lr = new LogRecorder(logRecorderName);
+        LogRecorder.Target target = new LogRecorder.Target(CleanLostNodesWork.class.getName(), Level.FINEST);
+        lr.setLoggers(List.of(target));
+        lrm.getRecorders().add(lr);
+    }
+
+    private static List<LogRecord> getLogRecords(JenkinsRule j, String logRecorderName) {
+        return j.jenkins.getLog().getLogRecorder(logRecorderName).getLogRecords();
```

Comment:
Beware that these are in a ring buffer, so if 256 messages are printed, the older ones will be lost. This could cause confusing test failures (or false positives).

Rather than relying on the Jenkins log recorders, it may be clearer and more robust to use our own `List<LogRecord>`, just like `LoggerRule` does. The tricky bit is figuring out where to keep such state inside the Jenkins JVM. I guess it could just be a `static` field in this class, since each Jenkins JVM would have its own copy of the class.

Reply:
noted about the ring buffer.
fixed now with using a static list and attaching a `logger.addHandler`, works. thank you